### PR TITLE
Translate h5py soft and hard linked datasets with an optional kwarg

### LINKED ISSUE #459: Kerchunk doesn't translate HDF5 hard links
I'm using kerchunk to translate hdf5 files in the [H5MD format ](https://www.nongnu.org/h5md/h5md.html) to be readable by zarr.
Kerchunk doesn't translate hard links- wherever in the directory the h5py dataset was last assigned is the only place it remains in the resulting zarr directory and all other hard links are not visible. 

Here's an example: First, I create an hdf5 file on the local filesystem
```python
import h5py
import numpy as np

file_name = "example.h5"
with h5py.File(file_name, "w") as f:
    data = np.arange(10)
    dset = f.create_dataset("original_dataset", data=data)

    f["linked_dataset"] = f["/original_dataset"]

with h5py.File(file_name, "r") as f:
    print(f["original_dataset"])
    print(f["linked_dataset"])
```

This gives the output:
```
<HDF5 dataset "original_dataset": shape (10,), type "<i8">
<HDF5 dataset "linked_dataset": shape (10,), type "<i8">
```
However, when I do:
```python
import kerchunk.hdf
import fsspec
import zarr

with fsspec.open(file_name) as inf:
    h5chunks = kerchunk.hdf.SingleHdf5ToZarr(
        inf, file_name, inline_threshold=100
    )
    fo = h5chunks.translate()

fs = fsspec.filesystem(
    "reference",
    fo=fo,
    skip_instance_cache=True,
)

z = zarr.open_group(fs.get_mapper(""), mode="r")
print(z.tree())
```
The tree looks like this:
```
/
 └── linked_dataset (10,) int64
```

This may be intentional behavior since zarr does not support linking datasets like hdf5 does. Is it possible to recreate the links in the json metadata created by `kerchunk.hdf.SingleHdf5ToZarr` to give the expected behavior? Please let me know if I'm missing anything!

For reference, I am using Kerchunk 0.2.5 and Python 3.11.9

### PR DESCRIPTION
**Pull request by ljwoods2**
Fixes #459 

Linked groups can't be translated as easily as linked datasets- the h5py method `visititems_links` is only called once for
each link. This means that if you have this hdf5 layout for example:
```
/
 └── particles
     └── trajectory
         ├── box
         │   └── edges
         │       ├── step (50,) int32
         │       ├── time (50,) float32
         │       └── value (50, 3, 3) float32
         └── position
             └── value (50, 13406, 3) float32
```
And you try to create this hard link and translate it:

```python
f["particles/trajectory2"] = f["particles/trajectory"]
```
Then `visititems_links` will be called on "box/edges/step", "time", and "value" during the traversal of "particles/trajectory" and therefore can't be traversed again during the traversal of "particles/trajectory2", resulting in an empty translated group. This behavior makes sense since it avoids the problem of circularly-linked groups, but it makes translating groups with h5py-builtin traversal methods difficult. If zarr-python 3.0 implements links, `visititems` and `visititems_links` may have to be replaced with custom directory traversal code as a result

This PR only allows translating linked datasets by traversing the h5py directory twice and using `require_dataset` instead of `create_dataset` to avoid duplicating non-linked groups and datasets.

### CODE REVIEW

**Comment by martindurant** [comment]
I'll look into the test failure, which looks unrelated. Maybe something about numpy 2, which just came out.

**Comment by martindurant** [comment]
(update: I have fixed one test failure so far, so one more to go)

**Comment by martindurant** [comment]
If you sync your branch, things should pass now.

**Comment by hmacdope** [comment]
@martindurant thanks for helping @ljwoods2 with this! This is part of his GSOC 2024 work with MDAnalysis, which I am co-supervising. With that hat on would you be able to approve the workflows here by any chance? Getting this merged would help us a lot moving forward.  Let me know if I can help in any way. 😄

**Comment by martindurant** [comment]
Sorry, new commits by themselves don't send a notification. Would you like me to add a label to this PR for gsoc?

File: kerchunk/hdf.py [hunk 1]
```diff
@@ -113,22 +113,39 @@ def __init__(
...
         dict
             Dictionary containing reference structure.
         """
         lggr.debug("Translation begins")
         self._transfer_attrs(self._h5f, self._zroot)
+
         self._h5f.visititems(self._translator)
+
+        if preserve_linked_dsets:
+            if h5py.__version__ < "3.11.0":
```
Comment (martindurant) [inline kerchunk/hdf.py:142]: Should we use more sophisticated version checking here? String comparisons are notoriously flaky.

File: kerchunk/hdf.py [hunk 2]
```diff
@@ -247,10 +264,23 @@ def _decode_filters(self, h5obj: Union[h5py.Dataset, h5py.Group]):
...
+        name: str,
+        h5obj: Union[
+            h5py.Dataset, h5py.Group, h5py.SoftLink, h5py.HardLink, h5py.ExternalLink
+        ],
+    ):
         """Produce Zarr metadata for all groups and datasets in the HDF5 file."""
         try:  # method must not raise exception
             kwargs = {}
+
+            if isinstance(h5obj, h5py.SoftLink) or isinstance(h5obj, h5py.HardLink):
```
Comment (martindurant) [inline kerchunk/hdf.py:278]:
```suggestion
            if isinstance(h5obj, h5py.SoftLink, h5py.HardLink)):
```

File: kerchunk/hdf.py [hunk 3]
```diff
@@ -247,10 +264,23 @@ def _decode_filters(self, h5obj: Union[h5py.Dataset, h5py.Group]):
...
+                h5obj = self._h5f[name]
+                if isinstance(h5obj, h5py.Group):
+                    # continues iteration of visititems_links
```
Comment (martindurant) [inline kerchunk/hdf.py:281]: logger.debug message?

**Comment by ljwoods2** [comment]
I switched out the string version compatibility with a direct check of if the `visititems_links` method is present- not sure if this is the most pythonic way to do it, let me know what you think. Also integrated your other suggestions

File: kerchunk/tests/test_hdf.py [hunk 4]
```diff
@@ -331,3 +334,38 @@ def test_inline_threshold():
...
     ).translate()
     assert inline_0 != inline_1_million
+
+
+def has_visititems_links():
+    try:
+        h5py.Group.visititems_links
+    except AttributeError:
+        return False
+    return True
```
Comment (martindurant) [inline kerchunk/tests/test_hdf.py:344]:
```suggestion
def has_visititems_links():
    return hasattr(h5py.Group, "visititems_links")
```

May as well use this in the main code too

File: kerchunk/hdf.py [hunk 5]
```diff
@@ -113,22 +114,41 @@ def __init__(
...
         self._h5f.visititems(self._translator)
+
+        if preserve_linked_dsets:
+            try:
+                h5py.Group.visititems_links
+            except AttributeError:
+                raise RuntimeError(
+                    "'preserve_linked_dsets' kwarg requires h5py 3.11.0 or later "
+                    f"is installed, found {version("h5py")}"
+                )
```
Comment (martindurant) [inline kerchunk/hdf.py:149]:
```suggestion
            except AttributeError:
                raise RuntimeError(
                    "'preserve_linked_dsets' kwarg requires h5py 3.11.0 or later "
                    f"is installed, found {h5py.__version__}"
                )
```

File: kerchunk/tests/test_hdf.py (hunk 4 above)
Comment (ljwoods2) [reply kerchunk/tests/test_hdf.py]: Good idea, thank you for the suggestions

File: kerchunk/hdf.py [hunk 6]
```diff
@@ -113,22 +114,39 @@ def __init__(
...
         self._transfer_attrs(self._h5f, self._zroot)
+
         self._h5f.visititems(self._translator)
+
+        if preserve_linked_dsets:
+            if not has_visititems_links():
+                raise RuntimeError(
+                    "'preserve_linked_dsets' kwarg requires h5py 3.11.0 or later "
+                    f"is installed, found {version("h5py")}"
+                )
```
Comment (martindurant) [inline kerchunk/hdf.py:147]:
```suggestion
                    "'preserve_linked_dsets' kwarg requires h5py 3.11.0 or later "
                    f"is installed, found {h5py.__version__)}"
                )
```

File: kerchunk/hdf.py [hunk 7]
```diff
@@ -113,22 +114,39 @@ def __init__(
...
         lggr.debug("Translation begins")
         self._transfer_attrs(self._h5f, self._zroot)
+
         self._h5f.visititems(self._translator)
+
+        if preserve_linked_dsets:
+            if not has_visititems_links():
+                raise RuntimeError(
+                    "'preserve_linked_dsets' kwarg requires h5py 3.11.0 or later "
+                    f"is installed, found {h5py.__version__)}"
```
Comment (martindurant) [inline kerchunk/hdf.py:146]:
```suggestion
                    f"is installed, found {h5py.__version__}"
```

File: kerchunk/tests/test_hdf.py [hunk 8]
```diff
@@ -331,3 +334,30 @@ def test_inline_threshold():
...
+    # 1. Test the hard linked datasets were translated correctly
+    # 2. Test the soft linked datasets were translated correctly
+    for link in ("hard", "soft"):
+        for dset in ("lat", "time"):
+            np.testing.assert_allclose(z[dset], z[f"{dset}_{link}"])
+            for key in z[f"{dset}_{link}"].attrs.keys():
+                if (
+                    key not in kerchunk.hdf._HIDDEN_ATTRS
+                    and key != "_ARRAY_DIMENSIONS"
+                ):
```
Comment (martindurant) [inline kerchunk/tests/test_hdf.py:362]:
```suggestion
                if key not in kerchunk.hdf._HIDDEN_ATTRS and key != "_ARRAY_DIMENSIONS":
```

File: kerchunk/hdf.py [hunk 9]
```diff
@@ -247,10 +265,24 @@ def _decode_filters(self, h5obj: Union[h5py.Dataset, h5py.Group]):
...
+    ):
         """Produce Zarr metadata for all groups and datasets in the HDF5 file."""
         try:  # method must not raise exception
             kwargs = {}
+
+            if isinstance(h5obj, (h5py.SoftLink, h5py.HardLink)):
+                h5obj = self._h5f[name]
+                if isinstance(h5obj, h5py.Group):
+                    # continues iteration of visititems_links
+                    lggr.debug(f"Skipping translation of HDF5 linked group: '{h5obj.name}'")
```
Comment (martindurant) [inline kerchunk/hdf.py:283]:
```suggestion
                    lggr.debug(
                        f"Skipping translation of HDF5 linked group: '{h5obj.name}'"
                    )
```

File: kerchunk/hdf.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 import io
 import logging
 from typing import Union, BinaryIO
+from importlib.metadata import version
```
Comment (martindurant) [inline kerchunk/hdf.py:5]:
```suggestion
```

File: kerchunk/hdf.py [hunk 11]
```diff
@@ -639,3 +672,6 @@ def _is_netcdf_datetime(dataset: h5py.Dataset):
 
 def _is_netcdf_variable(dataset: h5py.Dataset):
     return any("_Netcdf4" in _ for _ in dataset.attrs)
+
+def has_visititems_links():
```
Comment (martindurant) [inline kerchunk/hdf.py:676]:
```suggestion

def has_visititems_links():
```

**Comment by hmacdope** [comment]
Thanks a lot @martindurant! Appreciate your help here :+1:

**Pull request merged**